docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,446 @@
+
+ClusterAddonProvider for Helm could be implemented as a kubebuilder controller with a single reconciler watching for the Cluster object; each reconciliation will trigger Helm operations by leveraging on  https://github.com/helm/helm imported as a library.
+
+The ClusterAddonProvider for Helm is going to implement two new CRDs (see next paragraphs for the full spec). The first is named HelmChartProxy used to specify a Helm chart with configuration values and a selector identifying the clusters the add-on applies to. The second is called HelmReleaseProxy and contains information about the underlying Helm release installed on a cluster from a Helm chart. All the configurations will be done from the HelmChartProxy and the HelmReleaseProxy is simply used to maintain an inventory as well as surface information.
+
+This is a link to the prototype for [ClusterAddonProvider for Helm](https://github.com/Jont828/cluster-api-addon-provider-helm), and a demo can be found in this [Cluster API Office Hours recording](https://www.youtube.com/watch?v=dipR1Tzh4jE). The prototype implements HelmChartProxy and HelmReleaseProxy as well as controllers for both CRDs.
+
+#### Defining the list of Cluster add-ons to be installed
+
+ClusterAddonProvider for Helm will use HelmChartProxy to define the addons to install and which clusters they should be installed on. Each HelmChartProxy will contain fields to specify a single Helm chart. Additionally, the spec will contain a key value pair which will be used as a cluster selector The Helm chart is then installed on every cluster that includes a label matching the cluster selector and makes it part of the orchestration. If a selected cluster no longer matches the cluster selector label, then the Helm release will be uninstalled. 
+
+Note that the cluster selector label differs from the match label field in ClusterResourceSets as HelmChartProxy specifies a single key value pair whereas the match labels field matches any number of labels. The match labels field was not used because by adding another label pair, all of the currently matched clusters would be unselected and that would trigger the deletion of all associated Helm releases. 
+
+This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.
+
+#### Install and lifecycle manage the Package Manager
+
+Some Package Manager might require in-cluster components, like e.g. the kapp-controller, however this is not the case for Helm starting from v3, so this is a no-op.
+In case this is required, the ClusterAddonProvider should take care of this task as well.
```

Comment:
Sure thing. The way ClusterAddonProvider for Helm works is that it imports Helm v3 as a library in Go. Based on this design, if we wanted to upgrade to a future v4 of Helm, it would require releasing a new version of ClusterAddonProvider for Helm.

Reply:
Right, that makes sense. But I guess I was more talking about the generic `ClusterAddonProvider`. We are adding the tools installation (when necessary) to its responsibilities. Should we add upgrades as well?